ReactiveCocoa/Swift/Signal.swift

```diff
@@ -299,24 +299,94 @@ extension SignalType {
 	}
 }
 
-/// A reference type which wraps an array to avoid copying it for performance and
-/// memory usage optimization.
-private final class CollectState<Value> {
-	var values: [Value] = []
-
-	func append(value: Value) -> Self {
-		values.append(value)
-		return self
-	}
-}
-
 extension SignalType {
+
 	/// Returns a signal that will yield an array of values when `self` completes.
 	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
 	public func collect() -> Signal<[Value], Error> {
-		return self
-			.reduce(CollectState()) { $0.append($1) }
-			.map { $0.values }
+		return collect { _,_ in false }
+	}
+
+	/// Returns a signal that will yield an array of values until it reaches a certain
+	/// count.
+	///
+	/// When the count is reached the array is sent and the signal starts over 
+	/// yielding a new array of values.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func collect(count: Int) -> Signal<[Value], Error> {
+		return collect { values, _ in values.count == count }
+	}
+
+	/// Returns a signal that will yield an array of values based on a predicate.
+	///
+	/// The predicate should return `true` when the values should be sent and `false` 
+	/// when the values should be collected. The predicate receives the `values` 
+	/// already collected and the next value that should or should not be 
+	/// collected.
+	///
+	/// #### Example
+	///
+	///     let (signal, observer) = Signal<Int, NoError>.pipe()
+	///
+	///     signal
+	///         .collect { values, next in next != 7 }
+	///         .observeNext { print($0) }
+	///
+	///     observer.sendNext(1)
+	///     observer.sendNext(1)
+	///     observer.sendNext(7)
+	///     observer.sendNext(7)
+	///     observer.sendNext(5)
+	///     observer.sendNext(6)
+	///     observer.sendCompleted()
+	///
+	///     // Output:
+	///     // [1, 1]
+	///     // [7]
+	///     // [7]
+	///     // [5, 6]
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func collect(predicate: (values: [Value], next: Value) -> Bool) -> Signal<[Value], Error> {
```

Comment:
After thinking some more, I think it probably doesn't make sense to send `next` as an individual value. Instead, it could be included in `values`:

```swift
public fund collect(predicate: values: [Value] -> Bool) -> Signal<[Value], Error>
```

The `collect(count:)` above, for instance, checks `values.count == count`, but it's only going to be sent after another value should come in. If you pass 5 as the count, it's only going to send after the 6th value arrives (we should add a test for that). If the value was passed as part of `values`, it would correctly send after the 5th value.